asax/calculator.py

```diff
@@ -1,71 +1,118 @@
-import numpy as np
+from abc import ABC, abstractmethod
+from typing import Dict
 
+import numpy as np
+from jax.config import config
+from jax_md import space
+from ase.atoms import Atoms
 from ase.calculators.abc import GetPropertiesMixin
 from ase.calculators.calculator import compare_atoms, PropertyNotImplementedError
-from ase.constraints import full_3x3_to_voigt_6_stress
-
-from asax import utils
+from asax import utils, jax_utils
 
 
-class Calculator(GetPropertiesMixin):
-    def __init__(self, stress=False, x64=True):
-        self.x64 = x64
-        self.atoms = None
-        self.results = {}
+class Calculator(GetPropertiesMixin, ABC):
 
-        self.stress = stress
+    # TODO: Can this be abstract?
+    implemented_properties = ["energy", "forces"]
+    displacement: space.DisplacementFn
+    shift: space.ShiftFn
+    potential: jax_utils.PotentialFn
 
-        from jax.config import config
+    def __init__(self, x64=True):
+        self.x64 = x64
+        config.update("jax_enable_x64", self.x64)
 
-        config.update("jax_enable_x64", x64)
+        self.atoms_cache: Atoms = None
+        self.results = {}
 
-    def update(self, atoms):
-        if atoms is None and self.atoms is None:
+    def update(self, atoms: Atoms):
+        if atoms is None and self.atoms_cache is None:
             raise RuntimeError("Need an Atoms object to do anything!")
 
-        if self.atoms is None:
-            self.atoms = atoms.copy()
+        if self.atoms_cache is None:
```

Comment:
I simply renamed the property to make the sole purpose of caching more clear.

Reply:
Makes sense, but please revert it to `self.atoms` -- I don't see a reason to break with `ase` convention here. :) I'm not sure if doing it like this would break anything, but I can't 100% discount the possibility. Plus, `self.atoms` will definitely always exist, so no reason to introduce an additional property.